File: sp500_scanner/analysis/market_regime.py
"""Market regime detection — gates the scanner on overall market health.

Fetches SPY, QQQ, IWM, and ^VIX in a single batched yfinance call and scores
the market 0-100 across SPY trend, VIX fear, small-cap breadth (IWM), and
Nasdaq health (QQQ). main.py uses this score to decide whether to run the
full scan, watchlist-only, or abort before touching any S&P 500 ticker.
"""

import pandas as pd
import yfinance as yf

REGIME_TICKERS = ["SPY", "QQQ", "IWM", "^VIX"]
PERIOD = "1y"
INTERVAL = "1d"

# Fallback scores used when an instrument's data can't be fetched.
VIX_SCORE_DEFAULT = 12
BREADTH_SCORE_DEFAULT = 10
QQQ_SCORE_DEFAULT = 7

RUN_SCAN_THRESHOLD_DEFAULT = 55
ABORT_THRESHOLD = 40


class RegimeDataError(Exception):
    """Raised when SPY data can't be fetched — the scanner can't run without it."""


def _clean_close(df):
    """Return a cleaned close-price Series, or None if unusable."""
    if df is None or df.empty:
        return None

    df = df.rename(columns=str.lower)
    if "close" not in df.columns:
        return None

    close = df["close"].dropna()
    if close.empty:
        return None

    return close


def _fetch_closes():
    """Batch-download SPY/QQQ/IWM/^VIX and return {ticker: close_series or None}."""
    try:
        batch = yf.download(
            REGIME_TICKERS, period=PERIOD, interval=INTERVAL,
            group_by="ticker", auto_adjust=True, progress=False,
        )
    except Exception:
        batch = None

    closes = {}
    for ticker in REGIME_TICKERS:
        close = None
        if batch is not None and not batch.empty and isinstance(batch.columns, pd.MultiIndex):
            try:
                close = _clean_close(batch[ticker])
            except KeyError:
                close = None
        closes[ticker] = close

    return closes


def _safe_sma(close, window):
    """Rolling mean at `window`, falling back to the full-series mean if too short."""
    value = close.rolling(window).mean().iloc[-1]
    if pd.isna(value):
        value = close.mean()
    return value


def _score_spy_trend(close):
    price = close.iloc[-1]
    sma50 = _safe_sma(close, 50)
    sma200 = _safe_sma(close, 200)
    ema20 = close.ewm(span=20, adjust=False).mean().iloc[-1]

    above_200 = price > sma200
    above_50 = price > sma50
    golden_cross = sma50 > sma200
    above_ema20 = price > ema20

    score = 0
    if above_200:
        score += 15
    if above_50:
        score += 10
    if golden_cross:
        score += 10
    if above_ema20:
        score += 5

    if score >= 35:
        label = "STRONG BULL"
    elif score >= 25:
        label = "BULL"
    elif score >= 15:
        label = "NEUTRAL"
    elif score >= 5:
        label = "BEAR"
    else:
        label = "STRONG BEAR"

    return {
        "spy_price": price,
        "spy_sma50": sma50,
        "spy_sma200": sma200,
        "spy_ema20": ema20,
        "spy_above_200": above_200,
        "spy_above_50": above_50,
        "spy_above_ema20": above_ema20,
        "golden_cross": golden_cross,
        "spy_trend_score": score,
        "spy_trend_label": label,
    }


def _score_vix(close):
    vix_current = close.iloc[-1]
    vix_sma20 = _safe_sma(close, 20)

    if vix_current > 40:
        score = -10
        label = "CRISIS — exit everything"
    elif vix_current > 30:
        score = 0
        label = "PANIC — avoid new longs"
    elif vix_current > 25:
        score = 5
        label = "FEARFUL — reduce exposure"
    elif vix_current > 20:
        score = 12
        label = "CAUTIOUS — proceed carefully"
    elif vix_current > 15:
        score = 20
        label = "CALM — healthy market"
    else:
        score = 25
        label = "COMPLACENT — ideal bull conditions"

    trend = "RISING" if vix_current > vix_sma20 else "FALLING"

    return {
        "vix_current": vix_current,
        "vix_sma20": vix_sma20,
        "vix_trend": trend,
        "vix_score": score,
        "vix_label": label,
    }


def _score_breadth(close):
    price = close.iloc[-1]
    sma50 = _safe_sma(close, 50)
    sma200 = _safe_sma(close, 200)

    above_200 = price > sma200
    above_50 = price > sma50

    lookback = min(21, len(close) - 1)
    return_1m = (price / close.iloc[-1 - lookback] - 1) * 100 if lookback > 0 else 0.0

    score = 0
    if above_200:
        score += 10
    if above_50:
        score += 5
    if return_1m > 0:
        score += 5

    if score >= 15:
        label = "BROAD PARTICIPATION — best bull signal"
    elif score >= 8:
        label = "NARROW — large caps only leading"
    else:
        label = "WEAK BREADTH — distribution likely"

    return {
        "iwm_price": price,
        "iwm_sma50": sma50,
        "iwm_sma200": sma200,
        "iwm_above_200": above_200,
        "iwm_above_50": above_50,
        "iwm_return_1m": return_1m,
        "breadth_score": score,
        "breadth_label": label,
    }


def _score_qqq(close):
    price = close.iloc[-1]
    sma50 = _safe_sma(close, 50)
    sma200 = _safe_sma(close, 200)

    above_200 = price > sma200
    above_50 = price > sma50

    lookback = min(5, len(close) - 1)
    return_5d = (price / close.iloc[-1 - lookback] - 1) * 100 if lookback > 0 else 0.0

    score = 0
    if above_200:
        score += 8
    if above_50:
        score += 4
    if return_5d > 0:
        score += 3

    return {
        "qqq_price": price,
        "qqq_sma50": sma50,
        "qqq_sma200": sma200,
        "qqq_above_200": above_200,
        "qqq_above_50": above_50,
        "qqq_return_5d": return_5d,
        "qqq_score": score,
    }


def _classify_regime(score):
    if score >= 85:
        return "STRONG BULL", "✅✅"
    if score >= 70:
        return "BULL", "✅"
    if score >= 55:
        return "WEAK BULL", "⚠️"
    if score >= 40:
        return "NEUTRAL", "⚠️"
    if score >= 25:
        return "BEAR", "❌"
    return "STRONG BEAR", "❌❌"


def check_market_regime(threshold=RUN_SCAN_THRESHOLD_DEFAULT):
    """Fetch SPY/QQQ/IWM/VIX and compute the 0-100 market regime score.

    `threshold` is the minimum regime_score required to run the scan
    (default 55, overridable via --regime-threshold). The hard abort
    boundary (score < ABORT_THRESHOLD) is fixed regardless of `threshold`.

    Raises RegimeDataError if SPY data can't be fetched — the scanner
    cannot run without it.
    """
    closes = _fetch_closes()
    warnings = []

    spy_close = closes["SPY"]
    if spy_close is None:
        raise RegimeDataError("Failed to fetch SPY data — cannot run scanner without it")
    if len(spy_close) < 200:
        warnings.append(f"SPY history is only {len(spy_close)} bars (<200) — using available bars for long-term averages")
    spy = _score_spy_trend(spy_close)

    vix_close = closes["^VIX"]
    if vix_close is None:
        warnings.append("VIX fetch failed — using default VIX score (12/25)")
        vix = {
            "vix_current": 0.0,
            "vix_sma20": 0.0,
            "vix_trend": "UNKNOWN",
            "vix_score": VIX_SCORE_DEFAULT,
            "vix_label": "UNKNOWN — VIX data unavailable",
        }
    else:
        if len(vix_close) < 20:
            warnings.append(f"VIX history is only {len(vix_close)} bars (<20) — using available bars for trend average")
        vix = _score_vix(vix_close)

    iwm_close = closes["IWM"]
    if iwm_close is None:
        warnings.append("IWM fetch failed — using default breadth score (10/20)")
        breadth = {
            "iwm_price": 0.0,
            "iwm_sma50": 0.0,
            "iwm_sma200": 0.0,
            "iwm_above_200": False,
            "iwm_above_50": False,
            "iwm_return_1m": 0.0,
            "breadth_score": BREADTH_SCORE_DEFAULT,
            "breadth_label": "UNKNOWN — IWM data unavailable",
        }
    else:
        if len(iwm_close) < 200:
            warnings.append(f"IWM history is only {len(iwm_close)} bars (<200) — using available bars for long-term averages")
        breadth = _score_breadth(iwm_close)

    qqq_close = closes["QQQ"]
    if qqq_close is None:
        warnings.append("QQQ fetch failed — using default Nasdaq score (7/15)")
        qqq = {
            "qqq_price": 0.0,
            "qqq_sma50": 0.0,
            "qqq_sma200": 0.0,
            "qqq_above_200": False,
            "qqq_above_50": False,
            "qqq_return_5d": 0.0,
            "qqq_score": QQQ_SCORE_DEFAULT,
        }
    else:
        if len(qqq_close) < 200:
            warnings.append(f"QQQ history is only {len(qqq_close)} bars (<200) — using available bars for long-term averages")
        qqq = _score_qqq(qqq_close)

    regime_score = spy["spy_trend_score"] + vix["vix_score"] + breadth["breadth_score"] + qqq["qqq_score"]
    regime_score = max(0, min(100, regime_score))
    regime_label, regime_emoji = _classify_regime(regime_score)

    run_scan = regime_score >= threshold
    position_warning = threshold <= regime_score < 70
    vix_elevated = vix["vix_current"] > 30

    result = {
        "regime_score": regime_score,
        "regime_label": regime_label,
        "regime_emoji": regime_emoji,
        "run_scan": run_scan,
        "show_signals": run_scan,
        "position_warning": position_warning,
        "vix_elevated": vix_elevated,
        "warnings": warnings,
    }
    result.update(spy)
    result.update(vix)
    result.update(breadth)
    result.update(qqq)
    return result

File: src/utils/marketRegime.js
// Market regime detection — mirrors sp500_scanner/analysis/market_regime.py.
//
// Fetches SPY, QQQ, IWM, and ^VIX and scores the market 0-100 across SPY
// trend, VIX fear, small-cap breadth (IWM), and Nasdaq health (QQQ).

import { authHeaders } from './alpacaApi'
import { sma, ema } from './indicators'
import { isMajorEventThisWeek } from './economicCalendar'

const ALPACA_DATA_URL = 'https://data.alpaca.markets/v2/stocks'
export const ABORT_THRESHOLD = 40
export const RUN_SCAN_THRESHOLD_DEFAULT = 55

const VIX_SCORE_DEFAULT = 12
const BREADTH_SCORE_DEFAULT = 10
const QQQ_SCORE_DEFAULT = 7

export class RegimeDataError extends Error {}

function dateStr(d) {
  return d.toISOString().slice(0, 10)
}

export async function fetchAlpacaCloses(symbol) {
  const end = new Date()
  const start = new Date()
  start.setDate(start.getDate() - 400)

  const params = new URLSearchParams({
    timeframe: '1Day',
    start: dateStr(start),
    end: dateStr(end),
    limit: '1000',
    feed: 'iex',
    adjustment: 'split',
  })

  const response = await fetch(`${ALPACA_DATA_URL}/${symbol}/bars?${params}`, { headers: authHeaders() })
  if (!response.ok) {
    throw new Error(`Alpaca market data request failed (${response.status}) for ${symbol}`)
  }

  const data = await response.json()
  const bars = data.bars || []
  if (bars.length === 0) {
    throw new Error(`No market data returned for ${symbol}`)
  }
  return bars.map((b) => b.c)
}

// Rolling mean at `period`, falling back to the full-series mean if the
// series is shorter than `period` (mirrors _safe_sma's NaN fallback).
function safeSma(closes, period) {
  const value = sma(closes, period)
  if (value != null) return value
  return closes.reduce((a, b) => a + b, 0) / closes.length
}

// Generic 0-40 trend score (above 200MA +15, above 50MA +10, golden cross
// +10, above 20EMA +5), reused for SPY (market regime) and individual sector
// ETFs (sector regime).
export function scoreTrend(closes) {
  const price = closes[closes.length - 1]
  const sma50 = safeSma(closes, 50)
  const sma200 = safeSma(closes, 200)
  const ema20 = ema(closes, 20)
  const ema21 = ema(closes, 21) // display-only (e.g. "SPY vs 21 EMA") — scoring still uses the 20 EMA above

  const above200 = price > sma200
  const above50 = price > sma50
  const goldenCross = sma50 > sma200
  const aboveEma20 = price > ema20
  const above21 = ema21 != null && price > ema21

  let score = 0
  if (above200) score += 15
  if (above50) score += 10
  if (goldenCross) score += 10
  if (aboveEma20) score += 5

  let trendLabel
  if (score >= 35) trendLabel = 'STRONG BULL'
  else if (score >= 25) trendLabel = 'BULL'
  else if (score >= 15) trendLabel = 'NEUTRAL'
  else if (score >= 5) trendLabel = 'BEAR'
  else trendLabel = 'STRONG BEAR'

  return { price, sma50, sma200, ema20, ema21, above200, above50, above21, aboveEma20, goldenCross, trendScore: score, trendLabel }
}

// Percent return from `lookback` bars ago to the latest bar, clamped so it
// never looks further back than the series allows.
export function returnOverLookback(closes, lookback) {
  const n = closes.length
  const effectiveLookback = Math.min(lookback, n - 1)
  if (effectiveLookback <= 0) return 0
  const price = closes[n - 1]
  const past = closes[n - 1 - effectiveLookback]
  return (price / past - 1) * 100
}

function scoreSpyTrend(closes) {
  const t = scoreTrend(closes)
  return {
    spyPrice: t.price,
    spySma50: t.sma50,
    spySma200: t.sma200,
    spyEma20: t.ema20,
    spyEma21: t.ema21,
    spyAbove200: t.above200,
    spyAbove50: t.above50,
    spyAboveEma20: t.aboveEma20,
    spyAbove21: t.above21,
    goldenCross: t.goldenCross,
    spyTrendScore: t.trendScore,
    spyTrendLabel: t.trendLabel,
  }
}

function scoreVix(closes) {
  const vixCurrent = closes[closes.length - 1]
  const vixSma20 = safeSma(closes, 20)

  let vixScore
  let vixLabel
  if (vixCurrent > 40) {
    vixScore = -10
    vixLabel = 'CRISIS — exit everything'
  } else if (vixCurrent > 30) {
    vixScore = 0
    vixLabel = 'PANIC — avoid new longs'
  } else if (vixCurrent > 25) {
    vixScore = 5
    vixLabel = 'FEARFUL — reduce exposure'
  } else if (vixCurrent > 20) {
    vixScore = 12
    vixLabel = 'CAUTIOUS — proceed carefully'
  } else if (vixCurrent > 15) {
    vixScore = 20
    vixLabel = 'CALM — healthy market'
  } else {
    vixScore = 25
    vixLabel = 'COMPLACENT — ideal bull conditions'
  }

  const vixTrend = vixCurrent > vixSma20 ? 'RISING' : 'FALLING'

  return { vixCurrent, vixSma20, vixTrend, vixScore, vixLabel }
}

function scoreBreadth(closes) {
  const price = closes[closes.length - 1]
  const iwmSma50 = safeSma(closes, 50)
  const iwmSma200 = safeSma(closes, 200)

  const iwmAbove200 = price > iwmSma200
  const iwmAbove50 = price > iwmSma50
  const iwmReturn1m = returnOverLookback(closes, 21)

  let score = 0
  if (iwmAbove200) score += 10
  if (iwmAbove50) score += 5
  if (iwmReturn1m > 0) score += 5

  let breadthLabel
  if (score >= 15) breadthLabel = 'BROAD PARTICIPATION — best bull signal'
  else if (score >= 8) breadthLabel = 'NARROW — large caps only leading'
  else breadthLabel = 'WEAK BREADTH — distribution likely'

  return {
    iwmPrice: price,
    iwmSma50,
    iwmSma200,
    iwmAbove200,
    iwmAbove50,
    iwmReturn1m,
    breadthScore: score,
    breadthLabel,
  }
}

function scoreQqq(closes) {
  const price = closes[closes.length - 1]
  const qqqSma50 = safeSma(closes, 50)
  const qqqSma200 = safeSma(closes, 200)

  const qqqAbove200 = price > qqqSma200
  const qqqAbove50 = price > qqqSma50
  const qqqReturn5d = returnOverLookback(closes, 5)

  let score = 0
  if (qqqAbove200) score += 8
  if (qqqAbove50) score += 4
  if (qqqReturn5d > 0) score += 3

  return {
    qqqPrice: price,
    qqqSma50,
    qqqSma200,
    qqqAbove200,
    qqqAbove50,
    qqqReturn5d,
    qqqScore: score,
  }
}

export function classifyRegime(score) {
  if (score >= 85) return ['STRONG BULL', '✅✅']
  if (score >= 70) return ['BULL', '✅']
  if (score >= 55) return ['WEAK BULL', '⚠️']
  if (score >= 40) return ['NEUTRAL', '⚠️']
  if (score >= 25) return ['BEAR', '❌']
  return ['STRONG BEAR', '❌❌']
}

// Collapses the 0-100 regime score into the three position-sizing buckets
// the trading rules key off of: Risk On (full 1.5% risk), Risk Neutral
// (half size, 0.75%), Risk Off (cash, 0%). A major macro release this week
// downgrades an otherwise-Risk-On score to Risk Neutral.
export function classifyRiskEnvironment(regimeScore, date = new Date()) {
  if (regimeScore < ABORT_THRESHOLD) return 'off'
  if (regimeScore < 70) return 'neutral'
  if (isMajorEventThisWeek(date)) return 'neutral'
  return 'on'
}

// Fetch SPY/QQQ/IWM/^VIX and compute the 0-100 market regime score.
//
// `threshold` is the minimum regimeScore required to run the scan / show
// signals (default 55). The hard abort boundary (score < ABORT_THRESHOLD)
// is fixed regardless of `threshold`.
//
// Throws RegimeDataError if SPY data can't be fetched — the dashboard
// can't render without it.
export async function checkMarketRegime(threshold = RUN_SCAN_THRESHOLD_DEFAULT) {
  const warnings = []

  const [spyResult, iwmResult, qqqResult] = await Promise.allSettled([
    fetchAlpacaCloses('SPY'),
    fetchAlpacaCloses('IWM'),
    fetchAlpacaCloses('QQQ'),
  ])

  if (spyResult.status === 'rejected') {
    throw new RegimeDataError('Failed to fetch SPY data — cannot check market regime')
  }
  const spyCloses = spyResult.value
  if (spyCloses.length < 200) {
    warnings.push(`SPY history is only ${spyCloses.length} bars (<200) — using available bars for long-term averages`)
  }
  const spy = scoreSpyTrend(spyCloses)

  const vix = {
    vixCurrent: 0,
    vixSma20: 0,
    vixTrend: 'UNKNOWN',
    vixScore: VIX_SCORE_DEFAULT,
    vixLabel: 'UNKNOWN — VIX data unavailable',
  }

  let breadth
  if (iwmResult.status === 'rejected') {
    warnings.push('IWM fetch failed — using default breadth score (10/20)')
    breadth = {
      iwmPrice: 0,
      iwmSma50: 0,
      iwmSma200: 0,
      iwmAbove200: false,
      iwmAbove50: false,
      iwmReturn1m: 0,
      breadthScore: BREADTH_SCORE_DEFAULT,
      breadthLabel: 'UNKNOWN — IWM data unavailable',
    }
  } else {
    const iwmCloses = iwmResult.value
    if (iwmCloses.length < 200) {
      warnings.push(`IWM history is only ${iwmCloses.length} bars (<200) — using available bars for long-term averages`)
    }
    breadth = scoreBreadth(iwmCloses)
  }

  let qqq
  if (qqqResult.status === 'rejected') {
    warnings.push('QQQ fetch failed — using default Nasdaq score (7/15)')
    qqq = {
      qqqPrice: 0,
      qqqSma50: 0,
      qqqSma200: 0,
      qqqAbove200: false,
      qqqAbove50: false,
      qqqReturn5d: 0,
      qqqScore: QQQ_SCORE_DEFAULT,
    }
  } else {
    const qqqCloses = qqqResult.value
    if (qqqCloses.length < 200) {
      warnings.push(`QQQ history is only ${qqqCloses.length} bars (<200) — using available bars for long-term averages`)
    }
    qqq = scoreQqq(qqqCloses)
  }

  let regimeScore = spy.spyTrendScore + vix.vixScore + breadth.breadthScore + qqq.qqqScore
  regimeScore = Math.max(0, Math.min(100, regimeScore))
  const [regimeLabel, regimeEmoji] = classifyRegime(regimeScore)

  const runScan = regimeScore >= threshold
  const showSignals = runScan
  const positionWarning = regimeScore >= threshold && regimeScore < 70
  const vixElevated = vix.vixCurrent > 30

  return {
    regimeScore,
    regimeLabel,
    regimeEmoji,
    runScan,
    showSignals,
    positionWarning,
    vixElevated,
    warnings,
    ...spy,
    ...vix,
    ...breadth,
    ...qqq,
  }
}
